### PR TITLE
MXNet Neural Network performance improvements (reduced minibatch runtime on large datasets)

### PR DESCRIPTION
**Pull request by sgdread**
*Issue #, if available:*

*Description of changes:*

`nd.sum(loss).asscalar()` is doing implicit conversion tonumpy(), which is expensive (especially on GPU). Computing everything in NDArray speeds up computation. This problem exists in other parts of the NN training loop (scoring), but not as severe since loss is computed on batch level.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-174-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-174/1/index.html

**Comment by jwmueller** [comment]
Nice find!  Can you make the analogous changes in the training function in `tabular_nn_trial.py` which is used for HPO: 
 https://github.com/awslabs/autogluon/blob/master/autogluon/utils/tabular/ml/models/tabular_nn/tabular_nn_trial.py  

And then verify there are no bugs and the speedup persists with `hyperparameter_tune = True`?
To do this easily, you can run:  `python tests/unittests/test_tabularHPO.py`

**Comment by szha** [comment]
Job PR-174-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-174/2/index.html

**Comment by sgdread** [comment]
I updated HPO trials too. However, in the fast mode the number of batches in trials is very low and the improvement effects are not visible clearly. The speedup is mostly noticable in tabular data with lots of rows. Training 3 iterations on the following dataset: 582348 examples, 6 features (4 vector, 2 embedding, 0 language) went from 83s to 65s.

**Comment by szha** [comment]
Job PR-174-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-174/4/index.html

**Comment by zhanghang1989** [comment]
@jwmueller Does the code look good to you?

**Comment by szha** [comment]
Job PR-174-8 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-174/8/index.html

**Comment by jwmueller** [comment]
> @jwmueller Does the code look good to you?

yes am discussing it with Alex offline, will merge when confirmed that it won't cause any potential issues.

**Comment by Innixma** [comment]
Impacts of this change on a few problems (20 n_repeat 10 fold (200 NN models trained):

1630s -> 1408s
1667s -> 1459s
289s -> 275s

Improvements are stable and consistent across datasets. No issues have been found on any dataset.
